### PR TITLE
Extend community model

### PR DESCRIPTION
**Pull request by stelmo**
Currently our community model framework is rather fragile. Let's extend it to:

- [x] Create a community model type
- [x] List the joined metabolites and exchange reactions sensibly
- [x] Make it extensible, so any model type can be joined into a community

### CODE REVIEW

File: src/types/models/CommunityModel.jl [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+$(TYPEDSIGNATURES)
+
+Return the reactions in `cm`, which is a [`CommunityModel`](@ref). All reactions
+have the `id` of each respective underlying [`CommunityMember`](@ref) appended
+as a suffix. All exchange environmental reactions have `ENV_` added as a prefix
+followed by the metabolite id.
+"""
+function Accessors.reactions(cm::CommunityModel)
+    rxns = [rid * "#" * m.id for m in cm.members for rid in reactions(m.model)]
+    env_exs = ["ENV_" * env_met for env_met in get_env_mets(cm)]
```
Comment (exaexa) [inline src/types/models/CommunityModel.jl:65]: stringology should be parametrized (move the function into `constants` or something)

File: src/types/models/CommunityModel.jl [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    return vec
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Coupling constraint matrix for a [`CommunityModel`](@ref).
+"""
+Accessors.coupling(cm::CommunityModel) =
+    blockdiag([coupling(m.model) for m in cm.members]...)
```
Comment (exaexa) [inline src/types/models/CommunityModel.jl:190]: is this sparse?

File: test/types/temp_comm.jl [hunk 3]
```diff
@@ -0,0 +1,86 @@
+using COBREXA.Reconstruction
```
Comment (exaexa) [inline test/types/temp_comm.jl:1]: pls wrap this into a testset, also if possible keep all the imports in `runtests.jl`

File: src/types/models/CommunityModel.jl (hunk 2 above)
Comment (stelmo) [reply src/types/models/CommunityModel.jl:190]: It should be

File: test/types/temp_comm.jl (hunk 3 above)
Comment (stelmo) [reply test/types/temp_comm.jl:1]: will do

File: src/io/show/BalancedGrowthCommunityModel.jl [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        io,
+        "  and $(length(cm.exchange_reaction_ids)) exchange reactions that will be connected to the environment.",
+    )
+end
+
+function Base.show(io::Base.IO, ::MIME"text/plain", cm::BalancedGrowthCommunityModel)
+    println(io, "A balanced growth community model with:")
+    println(io, "  $(length(cm.members)) underlying models,")
+    println(io, "  objective $(cm.objective_id),")
+    if isempty(cm.env_met_flux_bounds)
```
Comment (exaexa) [inline src/io/show/BalancedGrowthCommunityModel.jl:20]: avoid being too verbose (we killed that for normal MetabolicModel and I don't see why it wouldn't be enough here)

File: src/io/show/BalancedGrowthCommunityModel.jl [hunk 5]
```diff
@@ -0,0 +1,28 @@
+function Base.show(io::Base.IO, ::MIME"text/plain", cm::CommunityMember)
```
Comment (exaexa) [inline src/io/show/BalancedGrowthCommunityModel.jl:1]: keep stuff single-line so it formats well in vectors etc

File: test/types/temp_comm.jl (hunk 3 above)
Comment (exaexa) [reply test/types/temp_comm.jl:1]: not resolved yet

File: test/types/temp_comm.jl [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+n_metabolites(cm)
+n_genes(cm)
+
+stoichiometry(cm)
+bounds(cm)
+objective(cm)
+
+
+
+d = flux_balance_analysis_dict(cm, Tulip.Optimizer)
```
Comment (exaexa) [inline test/types/temp_comm.jl:86]: remove

File: test/types/temp_comm.jl (hunk 3 above)
Comment (stelmo) [reply test/types/temp_comm.jl:1]: oh,   I forgot about it!

File: src/io/show/BalancedGrowthCommunityModel.jl (hunk 4 above)
Comment (stelmo) [reply src/io/show/BalancedGrowthCommunityModel.jl:20]: okay, I removed most of the words. But this then hides a lot of the internal structure... but I suppose if the person made the model, then they should know what is inside it!

File: src/io/show/BalancedGrowthCommunityModel.jl [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+        "A $(typeof(cm.model)) community member with $(n_reactions(cm.model)) reactions, $(n_metabolites(cm.model)) metabolites, and abundance $(cm.abundance*100)%.",
+    )
+end
+
+function Base.show(io::Base.IO, ::MIME"text/plain", cm::BalancedGrowthCommunityModel)
+    println(
+        io,
+        "A balanced growth community model comprised of $(length(cm.members)) underlying models.",
+    )
+end
```
Comment (exaexa) [inline src/io/show/BalancedGrowthCommunityModel.jl:13]: this isn't very useful, I guess the generic show will suffice for now. I'll eventually make a bigger PR with `MultiModelWrapper` that fixes this forever

Comment (exaexa) [reply src/io/show/BalancedGrowthCommunityModel.jl:13]: (as in, pls remove the custom `show`s here)

File: src/types/models/BalancedGrowthCommunityModel.jl [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+1. Exchange reactions, *all* of which are idenitified in `exchange_reaction_ids`,
+   have the form: `A[external] ⟷ ∅` where `A` is a metabolite. No other
+   exchanges are allowed.
+2. The biomass reaction of a model produces a biomass metabolite called
+   `biomass_metabolite_id` with stoichiometry 1.
+3. There is only one biomass reaction in the model.
+"""
+Base.@kwdef mutable struct CommunityMember
+    "Name of model appended to intracellular reactions and metabolites."
+    id::String
```
Comment (exaexa) [inline src/types/models/BalancedGrowthCommunityModel.jl:20]: TODO for myself, for later:
`id` should be outside of the community member, making the `members` a dict.

File: src/types/models/BalancedGrowthCommunityModel.jl [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    "Abundance of organism."
+    abundance::Float64
+    "Underlying model."
+    model::AbstractMetabolicModel
+    "List of all exchange reactions in model."
+    exchange_reaction_ids::Vector{String}
+    "ID of biomass metabolite."
+    biomass_metabolite_id::String
```
Comment (exaexa) [inline src/types/models/BalancedGrowthCommunityModel.jl:28]: TODO for myself later: biomass balancing is a function of the class that provides the balance, so the metabolite (= equality constraint) should be created externally, not by the community member.

**Review by exaexa — APPROVED** [review]
Let's get this in, I added some TODOs for myself for later.

(I'll convert them to issues)

**Pull request merged**